File: agent/crypto/src/replay_guard.cpp
#include "jd_relay/crypto/replay_guard.h"
#include <chrono>
#include <algorithm>

namespace jd_relay::crypto {

ReplayGuard::ReplayGuard(int window_seconds)
    : window_seconds_(window_seconds) {}

int64_t ReplayGuard::now_ms() {
    auto now = std::chrono::system_clock::now();
    return std::chrono::duration_cast<std::chrono::milliseconds>(
        now.time_since_epoch()).count();
}

bool ReplayGuard::is_within_window(int64_t timestamp) const {
    int64_t now = now_ms();
    int64_t window_ms = static_cast<int64_t>(window_seconds_) * 1000;
    // Check timestamp is within [now - window, now + window]
    // We allow a small future skew (±window) to handle clock drift
    return (timestamp >= now - window_ms) && (timestamp <= now + window_ms);
}

bool ReplayGuard::is_replay(const std::string& nonce) const {
    return nonces_.find(nonce) != nonces_.end();
}

void ReplayGuard::record_nonce(const std::string& nonce) {
    int64_t expiry = now_ms() + static_cast<int64_t>(window_seconds_) * 1000;
    nonces_[nonce] = expiry;
}

bool ReplayGuard::check_and_record(int64_t timestamp, const std::string& nonce) {
    // 1. Check timestamp window
    if (!is_within_window(timestamp)) {
        return false;
    }

    // 2. Check nonce uniqueness
    if (is_replay(nonce)) {
        return false;
    }

    // 3. Record nonce
    record_nonce(nonce);

    // 4. Purge expired nonces periodically
    if (nonces_.size() > 10000) {
        purge_expired();
    }

    return true;
}

void ReplayGuard::purge_expired() {
    int64_t now = now_ms();
    for (auto it = nonces_.begin(); it != nonces_.end(); ) {
        if (it->second < now) {
            it = nonces_.erase(it);
        } else {
            ++it;
        }
    }
}

} // namespace jd_relay::crypto

File: forwarder/app/crypto/replay_guard.py
"""Anti-replay protection — wire-compatible with C++ ReplayGuard.

Uses timestamp window (±5 min default) + nonce cache to detect replays.
Validation order: timestamp → nonce uniqueness (matches C++ exactly).
"""

import time


class ReplayGuard:
    """Anti-replay protection using timestamp window + nonce cache.

    Matches C++ ReplayGuard behavior:
    - Timestamp window: ±window_seconds (default 300 = 5 minutes)
    - Nonce cache: nonce → expiry timestamp (ms)
    - Auto-purge when cache exceeds 10,000 entries
    """

    def __init__(self, window_seconds: int = 300):
        self._window_seconds = window_seconds
        self._nonces: dict[str, int] = {}  # nonce → expiry timestamp (ms)

    @staticmethod
    def now_ms() -> int:
        """Get current time as Unix milliseconds."""
        return int(time.time() * 1000)

    def is_within_window(self, timestamp: int) -> bool:
        """Check if timestamp is within the acceptable window."""
        now = self.now_ms()
        window_ms = self._window_seconds * 1000
        return (timestamp >= now - window_ms) and (timestamp <= now + window_ms)

    def is_replay(self, nonce: str) -> bool:
        """Check if a nonce has already been seen."""
        return nonce in self._nonces

    def record_nonce(self, nonce: str) -> None:
        """Add a nonce to the cache with an expiry timestamp."""
        expiry = self.now_ms() + self._window_seconds * 1000
        self._nonces[nonce] = expiry

    def check_and_record(self, timestamp: int, nonce: str) -> bool:
        """Check timestamp window + nonce uniqueness, record if fresh.

        Returns True if message is fresh (within window AND nonce not seen).
        On success, the nonce is recorded. Auto-purges when > 10k entries.
        """
        if not self.is_within_window(timestamp):
            return False
        if self.is_replay(nonce):
            return False

        self.record_nonce(nonce)

        if len(self._nonces) > 10000:
            self.purge_expired()

        return True

    def purge_expired(self) -> None:
        """Remove all nonces whose expiry time has passed."""
        now = self.now_ms()
        self._nonces = {
            nonce: expiry
            for nonce, expiry in self._nonces.items()
            if expiry >= now
        }

    @property
    def cache_size(self) -> int:
        """Number of nonces currently cached."""
        return len(self._nonces)
